lib/index.js

```diff
@@ -0,0 +1,144 @@
+var hl = require('highland');
+var Promise = require('bluebird');
+var redis = require('then-redis');
+var retry = require('retry');
+var stream = require('stream');
+
+function checkpoint(options) {
+    if (!options) throw new Error("You must supply options");
+    this.messageLimit = options.messages || 3;
+    this.redisClient = this.redisClient || redis.createClient(this.redisUrl);
+    return this;
+}
+
+function consume(streamConstructor) {
+    if (!(streamConstructor instanceof Function)) throw new Error('Module requires function');
+
+    var consumer = this;
+    var dataBuf = '';
+    var event = '';
+    var id = '';
+
+    this.stream = getCheckpointId(streamConstructor)
+        .spread(createHighlandStream)
+        .then(initiateStream);
+    return consumer;
+
+    function getCheckpointId(constructor) {
+        if (consumer.redisClient) {
+            return consumer.redisClient.get('checkpoint')
+                .then(function(checkpoint) {
+                    var id = checkpoint ? JSON.parse(checkpoint).id : undefined;
+                    return Promise.resolve([id, constructor]);
+                });
+        }
+        return Promise.resolve([undefined, constructor]);
+    }
+
+    function createHighlandStream(id, streamConstructor) {
+        var tryStream;
+        try {
+            tryStream = streamConstructor(id);
+        } catch(err) {
+            throw new Error("Could not establish SSE stream", err);
+        }
+
+        if (tryStream instanceof stream) return Promise.resolve(hl(tryStream));
+        if (typeof tryStream.then === 'function') {
+            return tryStream.then(function(foo) {
+                return Promise.resolve(hl(foo));
+            });
+        }
+        throw new Error("Constructor did not return a readable stream");
+    }
+
+    function initiateStream(stream) {
+        this.stream = stream;
+        this.stream.resume();
+
+        return this.stream
+            .through(hl.consume(parseJSON))
+            .through(hl.consume(tryCallback))
+            .each(createRedis);
+    }
+
+    function parseJSON(err, line, push, next) {
+        var line = line.toString().replace('\n', '');
+        if (line === '') {
+            if (!dataBuf) return;
+            var data = dataBuf;
+            dataBuf = '';
+            push(null, {
+                id : id,
+                event : event,
+                data : data
+            });
+            return next()
+        }
+        if (line.indexOf('data: ') === 0) dataBuf += (dataBuf ? '\n' : '') + line.slice(6);
+        if (line.indexOf('id: ') === 0) id = line.slice(4);
+        if (line.indexOf('event: ') === 0) event = line.slice(6);
+        next();
+    }
+
+    function tryCallback(err, chunk, push, next) {
+        if (!consumer.eventHook) {
+            push(err, chunk);
+            return next();
+        }
+
+        var operation = retry.operation({
+            retries: 10000,
```

Comment:
Don't mind that we have the option of setting a max amount of attempts, however the default should be to retry forever given how we apply the SSE pattern overall with queuing

Setting the forever option is possible : https://github.com/tim-kos/node-retry/blob/d0b863bb87b486a77eec20d6ec3f5fd2f35c6771/test/integration/test-retry-operation.js#L84

We could add in another ticket to change the signature a bit, would allows the user of the library to pass in the exact option hash for retry?

```js
var retryOptions = {
  retries: 5,
  factor: 3,
  minTimeout: 1 * 1000,
  maxTimeout: 60 * 1000,
  randomize: true,
}
consumer(retryOptions)
    .consume(function makeSSEStream(lastEventId) {
        return grabToken().then(function(token) {
            return request({
                uri: 'http://myapi.example.com/changes/stream?resources=trackingData,equipment',
                headers: {'Authorization': 'Bearer ' + token, 'last-event-id': lastEventId}
            });
        })
    })
    .onEvent(function (sse) {
        return new Promise(function (resolve) {
            exchange.publish({text: sse.data}, {key: sse.event}).on('drain', resolve);
        })
    })
    .checkpoint({redisUrl: 'redis://someuser:secret@localhost:10242', time: '10s', messages: 5});
```